src/rsa.rs

```diff
@@ -163,8 +165,107 @@ fn parse_public_key<'a>(input: untrusted::Input<'a>) ->
     })
 }
 
+#[allow(non_camel_case_types)]
+enum BIGNUM {}
+#[allow(non_camel_case_types)]
+enum BN_MONT_CTX {}
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct RSA {
+    n: *mut BIGNUM,
+    e: *mut BIGNUM,
+    d: *mut BIGNUM,
+    p: *mut BIGNUM,
+    q: *mut BIGNUM,
+    dmp1: *mut BIGNUM,
+    dmq1: *mut BIGNUM,
+    iqmp: *mut BIGNUM,
+    mont_n: *mut BN_MONT_CTX,
+    mont_p: *mut BN_MONT_CTX,
+    mont_q: *mut BN_MONT_CTX,
+    mont_qq: *mut BN_MONT_CTX,
+    qmn_mont: *mut BIGNUM,
+    iqmp_mont: *mut BIGNUM,
+}
+
+/// An RSA key pair.
+pub struct RSAKeyPair {
+    rsa: Box<RSA>,
+}
+
+const BN_FLG_CONSTTIME: c::int = 4;
+
+fn build_bn(input: untrusted::Input) -> *mut BIGNUM {
+    unsafe {
+        let bytes = input.as_slice_less_safe();
+        let res = BN_bin2bn(bytes.as_ptr(), bytes.len(), ptr::null_mut());
+        BN_set_flags(res, BN_FLG_CONSTTIME);
+        res
+    }
+}
+
+impl RSAKeyPair {
+    /// Reads key data from the given DER-encoded input, which must contain
+    /// the full private key data (n, e, d, p, q, dmp1, dmq1, iqmp).
+    ///
+    /// This is intended for use by code that deserializes key pairs.
+    pub fn from_der(input: untrusted::Input) -> Result<RSAKeyPair, ()> {
+        input.read_all((), |input| {
+            der::nested(input, der::Tag::Sequence, (), |input| {
+                let _ = try!(der::small_nonnegative_integer(input)); // version
+                let n = try!(der::positive_integer(input));
+                let e = try!(der::positive_integer(input));
+                let d = try!(der::positive_integer(input));
+                let p = try!(der::positive_integer(input));
+                let q = try!(der::positive_integer(input));
+                let dmp1 = try!(der::positive_integer(input));
+                let dmq1 = try!(der::positive_integer(input));
+                let iqmp = try!(der::positive_integer(input));
+                let mut rsa = Box::new(RSA {
+                    n: build_bn(n), e: build_bn(e), d: build_bn(d),
+                    p: build_bn(p), q: build_bn(q), dmp1: build_bn(dmp1),
+                    dmq1: build_bn(dmq1), iqmp: build_bn(iqmp),
+                    mont_n: ptr::null_mut(), mont_p: ptr::null_mut(),
+                    mont_q: ptr::null_mut(), mont_qq: ptr::null_mut(),
+                    qmn_mont: ptr::null_mut(), iqmp_mont: ptr::null_mut(),
+                });
+                try!(bssl::map_result(unsafe {
+                    rsa_new_end(rsa.as_mut())
+                }));
+                Ok(RSAKeyPair { rsa: rsa })
+            })
+        })
+    }
+}
+
+impl Drop for RSAKeyPair {
+    fn drop(&mut self) {
```

Comment:
It's still used, in the error handling for `RSA_generate()`. So it should be removed as part of adding `RSAKeyPair` generation, I think.